packages/wallet-core/src/protocols/direct-funder.ts

```diff
@@ -0,0 +1,253 @@
+import * as _ from 'lodash';
+
+import {checkThat, isSimpleAllocation, unreachable} from '../utils';
+import {Message} from '../wire-protocol';
+import {BN} from '../bignumber';
+import {addHash, hashState, signState} from '../state-utils';
+import {Address, SignatureEntry, State, StateWithHash, Uint256} from '../types';
+
+type AddressedMessage = {recipient: string; message: Message};
+
+export type OpenChannelEvent =
+  | {type: 'MessageReceived'; message: Message}
+  | {type: 'FundingUpdated'; amount: Uint256; finalized: boolean};
+
+type SignedStateHash = {hash: string; signatures: SignatureEntry[]};
```

Comment:
Is this not already a type we have defined in wallet-core?

Reply:
I don't think so! We typically pass around states. In this work, I am suggesting a slightly clever form of compression:
1. the preFS and postFS states are the same, except for their turn number.
2. we can therefore compute the hashes up front, and

One benefit of this is that it enables some aggressive network optimization, where we can send
```
{ objectiveId: string; hashes: SignedStateHash[] }
```

across the wire, instead of this, which contains a lot more data.
```
{ signedStates: SignedState[] }
```

The first wire protocol also has an advantage in terms of security: messages for an unauthorized objective can be ignored.